File: mobile/expo-driver-app/src/services/location.service.ts
import * as Location from 'expo-location';
import * as TaskManager from 'expo-task-manager';
import apiService from './api.service';
import { API_ENDPOINTS, APP_CONFIG } from '../config/api';

const LOCATION_TASK_NAME = 'background-location-task';

class LocationService {
  private watchId: Location.LocationSubscription | null = null;
  private currentJobId: string | null = null;

  async requestPermissions(): Promise<boolean> {
    const { status: foregroundStatus } = await Location.requestForegroundPermissionsAsync();
    
    if (foregroundStatus !== 'granted') {
      console.log('❌ Foreground location permission denied');
      return false;
    }

    const { status: backgroundStatus } = await Location.requestBackgroundPermissionsAsync();
    
    if (backgroundStatus !== 'granted') {
      console.log('⚠️ Background location permission denied');
      // Still return true as foreground is granted
    }

    console.log('✅ Location permissions granted');
    return true;
  }

  async getCurrentLocation(): Promise<Location.LocationObject | null> {
    try {
      return await Location.getCurrentPositionAsync({
        accuracy: Location.Accuracy.High,
      });
    } catch (error) {
      console.error('Error getting location:', error);
      return null;
    }
  }

  async startTracking(jobId: string): Promise<void> {
    this.currentJobId = jobId;
    console.log('📍 Starting location tracking for job:', jobId);

    // Start foreground location updates
    this.watchId = await Location.watchPositionAsync(
      {
        accuracy: Location.Accuracy.High,
        distanceInterval: APP_CONFIG.LOCATION_DISTANCE_FILTER,
        timeInterval: APP_CONFIG.LOCATION_UPDATE_INTERVAL,
      },
      async (location) => {
        await this.sendLocationUpdate(location);
      }
    );

    // Start background location updates
    await Location.startLocationUpdatesAsync(LOCATION_TASK_NAME, {
      accuracy: Location.Accuracy.High,
      distanceInterval: APP_CONFIG.LOCATION_DISTANCE_FILTER,
      timeInterval: APP_CONFIG.LOCATION_UPDATE_INTERVAL,
      foregroundService: {
        notificationTitle: 'Speedy Van Driver',
        notificationBody: 'Tracking your location for active delivery',
      },
    });
  }

  async stopTracking(): Promise<void> {
    console.log('📍 Stopping location tracking');
    this.currentJobId = null;

    // Stop foreground updates
    if (this.watchId) {
      this.watchId.remove();
      this.watchId = null;
    }

    // Stop background updates
    const isRegistered = await TaskManager.isTaskRegisteredAsync(LOCATION_TASK_NAME);
    if (isRegistered) {
      await Location.stopLocationUpdatesAsync(LOCATION_TASK_NAME);
    }
  }

  async sendLocationUpdate(location: Location.LocationObject): Promise<void> {
    if (!this.currentJobId) return;

    try {
      await apiService.post(API_ENDPOINTS.SEND_LOCATION, {
        bookingId: this.currentJobId,
        latitude: location.coords.latitude,
        longitude: location.coords.longitude,
        accuracy: location.coords.accuracy,
      });
      console.log('✅ Location update sent');
    } catch (error) {
      console.error('❌ Failed to send location update:', error);
    }
  }
}

// Define background location task
TaskManager.defineTask(LOCATION_TASK_NAME, async ({ data, error }: any) => {
  if (error) {
    console.error('Background location error:', error);
    return;
  }

  if (data) {
    const { locations } = data;
    if (locations && locations.length > 0) {
      const location = locations[0];
      console.log('📍 Background location update:', location.coords);
      // Location updates are handled by the foreground watcher
    }
  }
});

export default new LocationService();

File: mobile/ios-driver-app/Services/LocationService.swift
import Foundation
import CoreLocation
import Combine

// MARK: - Location Service

class LocationService: NSObject, ObservableObject {
    static let shared = LocationService()
    
    @Published var currentLocation: CLLocation?
    @Published var authorizationStatus: CLAuthorizationStatus = .notDetermined
    @Published var isTracking = false
    
    private let locationManager = CLLocationManager()
    private let network = NetworkService.shared
    private var updateTimer: Timer?
    private var activeJobId: String?
    
    private override init() {
        super.init()
        setupLocationManager()
    }
    
    // MARK: - Setup
    
    private func setupLocationManager() {
        locationManager.delegate = self
        locationManager.desiredAccuracy = kCLLocationAccuracyBest
        locationManager.distanceFilter = AppConfig.minimumDistanceFilter
        locationManager.allowsBackgroundLocationUpdates = AppConfig.backgroundLocationEnabled
        locationManager.pausesLocationUpdatesAutomatically = false
        locationManager.showsBackgroundLocationIndicator = true
        
        // Check initial authorization status
        if #available(iOS 14.0, *) {
            authorizationStatus = locationManager.authorizationStatus
        } else {
            authorizationStatus = CLLocationManager.authorizationStatus()
        }
    }
    
    // MARK: - Permission Management
    
    func requestPermission() {
        print("📍 Requesting location permission...")
        
        switch authorizationStatus {
        case .notDetermined:
            locationManager.requestWhenInUseAuthorization()
            
        case .authorizedWhenInUse:
            // Request always authorization for background tracking
            locationManager.requestAlwaysAuthorization()
            
        case .denied, .restricted:
            print("⚠️ Location permission denied or restricted")
            
        case .authorizedAlways:
            print("✅ Location permission already granted (always)")
            
        @unknown default:
            break
        }
    }
    
    // MARK: - Tracking Control
    
    func startTracking(for jobId: String) {
        print("📍 Starting location tracking for job: \(jobId)")
        
        activeJobId = jobId
        isTracking = true
        
        // Request permission if needed
        if authorizationStatus == .notDetermined {
            requestPermission()
        }
        
        // Start location updates
        locationManager.startUpdatingLocation()
        
        // Start timer for periodic updates
        updateTimer = Timer.scheduledTimer(
            withTimeInterval: AppConfig.locationUpdateInterval,
            repeats: true
        ) { [weak self] _ in
            self?.sendLocationUpdate()
        }
    }
    
    func stopTracking() {
        print("📍 Stopping location tracking")
        
        isTracking = false
        activeJobId = nil
        
        locationManager.stopUpdatingLocation()
        updateTimer?.invalidate()
        updateTimer = nil
    }
    
    // MARK: - Location Updates
    
    private func sendLocationUpdate() {
        guard let location = currentLocation,
              let jobId = activeJobId else {
            print("⚠️ Cannot send location update: no location or job ID")
            return
        }
        
        print("📍 Sending location update...")
        
        Task {
            do {
                let request = LocationUpdateRequest(
                    bookingId: jobId,
                    latitude: location.coordinate.latitude,
                    longitude: location.coordinate.longitude,
                    accuracy: location.horizontalAccuracy
                )
                
                let response: LocationUpdateResponse = try await network.request(
                    .sendLocation,
                    method: .post,
                    body: request
                )
                
                if response.success {
                    print("✅ Location update sent successfully")
                } else {
                    print("⚠️ Location update failed: \(response.error ?? "Unknown error")")
                }
            } catch {
                print("❌ Failed to send location update: \(error.localizedDescription)")
            }
        }
    }
    
    // MARK: - Manual Location Update
    
    func sendManualUpdate(for jobId: String) {
        guard let location = currentLocation else { return }
        
        Task {
            do {
                let request = LocationUpdateRequest(
                    bookingId: jobId,
                    latitude: location.coordinate.latitude,
                    longitude: location.coordinate.longitude,
                    accuracy: location.horizontalAccuracy
                )
                
                let _: LocationUpdateResponse = try await network.request(
                    .sendLocation,
                    method: .post,
                    body: request
                )
            } catch {
                print("❌ Manual location update failed: \(error.localizedDescription)")
            }
        }
    }
}

// MARK: - CLLocationManagerDelegate

extension LocationService: CLLocationManagerDelegate {
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let location = locations.last else { return }
        
        currentLocation = location
        
        print("📍 Location updated: \(location.coordinate.latitude), \(location.coordinate.longitude)")
        
        // Send immediate update if tracking
        if isTracking {
            sendLocationUpdate()
        }
    }
    
    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        print("❌ Location error: \(error.localizedDescription)")
    }
    
    func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        if #available(iOS 14.0, *) {
            authorizationStatus = manager.authorizationStatus
        } else {
            authorizationStatus = CLLocationManager.authorizationStatus()
        }
        
        print("📍 Location authorization changed: \(authorizationStatus.rawValue)")
        
        // Start tracking if authorized and tracking was requested
        if authorizationStatus == .authorizedAlways || authorizationStatus == .authorizedWhenInUse {
            if isTracking {
                locationManager.startUpdatingLocation()
            }
        }
    }
}

